mgr/cephadm: unset MON flag during upgrade

shraddhaag opened the pull request:
This PR unsets the specified mon flags before an upgrade starts and sets them after an upgrade. This will be used to disable availability tracking before an upgrade and enable it again after the upgrade is complete. The code is written in an extensible way, such that any further usecases will only have to make a single line change to achieve the same. 

Scenarios handled: 


|Type|Resolution|
|-----|-----------|
| Full upgrade | flag unset before upgrade; flag reset after upgrade. |
|Staggered upgrade| flag unset and reset for each upgrade step triggered by user.|
|Upgrade stopped by user manually using `ceph orch upgrade stop`| flag is reset before command exits|
|Upgrade fails| flag is reset before command exits|

Fixes: https://tracker.ceph.com/issues/72489

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

shraddhaag committed:
mgr/cephadm: unset MON flag during upgrade

This commit unsets the specified mon flags before an upgrade
starts and sets them after an upgrade.

For full upgrade, it unsets the flag before upgrade and resets
it after upgrade is complete.

For staggered upgrade, it will unset and set the flag for each
section of upgrade done separately.

If upgrade is stopped by the user, the flag will be reset before
upgrade stops.

If the upgrade fails, the flag will be reset before upgrade
exists.

Fixes: https://tracker.ceph.com/issues/72489
Signed-off-by: Shraddha Agrawal <shraddhaag@ibm.com>

Signed-off-by: Shraddha Agrawal <shraddha.agrawal000@gmail.com>

adk3798 commented on src/pybind/mgr/orchestrator/module.py:
```diff
@@ -2515,7 +2515,8 @@ def _upgrade_check_image_name(image: Optional[str], ceph_version: Optional[str])
                        image: Optional[str] = None,
-                       ceph_version: Optional[str] = None) -> HandleCommandResult:
+                       ceph_version: Optional[str] = None,
+                       no_mon_flags: bool = False) -> HandleCommandResult:
```
Is this meant to be on the `upgrade start` command rather than `upgrade check`? It doesn't looks like upgrade check would have anything to do with mon flags

shraddhaag replied:
You're right, addressed!

adk3798 commented on src/pybind/mgr/orchestrator/module.py:
```diff
@@ -2569,7 +2570,7 @@ def _upgrade_start(self,
         dtypes = daemon_types.split(',') if daemon_types is not None else None
         service_names = services.split(',') if services is not None else None
-        completion = self.upgrade_start(image, ceph_version, dtypes, hosts, service_names, limit)
+        completion = self.upgrade_start(image, ceph_version, dtypes, hosts, service_names, limit, no_mon_flags)
```
in relation to previous comment, I think `no_mon_flags` would be undefined here

adk3798 commented on src/pybind/mgr/orchestrator/_interface.py:
```diff
@@ -987,7 +987,7 @@ def upgrade_ls(self, image: Optional[str], tags: bool, show_all_versions: Option
 
     def upgrade_start(self, image: Optional[str], version: Optional[str], daemon_types: Optional[List[str]],
-                      hosts: Optional[str], services: Optional[List[str]], limit: Optional[int]) -> OrchResult[str]:
+                      hosts: Optional[str], services: Optional[List[str]], limit: Optional[int], set_mon_flags: bool = False) -> OrchResult[str]:
```
This seems to be the only place we're using the positive "set mon flags" and then everywhere else it's the negative "no mon flags". Would it be possible to keep using the positive statement through the full path from here to the actual upgrade code?

shraddhaag replied:
Using `unset_mon_flags` throughout so there is no discrepancy.

adk3798 commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -1328,5 +1351,18 @@ def _do_upgrade(self):
             })
 
+        # Upgrade is complete. Unset mon flags
+        if self.upgrade_state.mon_flags:
```
this handles the successfully completed upgrade case, but I think we additionally need to handle the case where users manually stop the upgrade before it's completed.

We should also consider the staggered upgrade scenario. If a user does a staggered upgrade, say for just the mgr daemons, would we want to unset the flags when that completes and then set them again when they trigger an additional staggered upgrade for the next daemon type/service/host etc.? It doesn't feel totally correct, but it might be difficult to do it any other way (as in, "knowing" there will be additional staggered upgrades triggered and not unsetting the flags) so we might have to handle it like that.

shraddhaag replied:
I've handled full upgrades, staggered upgrades, manual upgrade interruptions and upgrade failures. Could you please review this again and see if it suffices?

shraddhaag committed:
mgr/cephadm/module.py: unset enable_availability_tracking on upgrades

This commit ensures that availability tracking is automatically
disabled at the time of upgrade start and disabled at the end
of an upgrade. If the feature is not enabled in the first place,
upgrade would not toggle the config flag.

Fixes: https://tracker.ceph.com/issues/72489
Signed-off-by: Shraddha Agrawal <shraddhaag@ibm.com>
Signed-off-by: Shraddha Agrawal <shraddha.agrawal000@gmail.com>

adk3798 commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -1176,6 +1199,27 @@ def _do_upgrade(self):
+                # set mon_flags to the list of flags actually unset before upgrade.
+                # this is to ensure we only re-set these flags at end of upgrade 
+                # and don't mistakenly set any flag that wasn't set before upgrade. 
+                self.upgrade_state.mon_flags = flags_toggled
```
there might be a slight issue with doing this here. In a single upgrade I'd expect `_do_upgrade` to run a number of times. It looks like we're building the list of flags stored in the `mon_flags` attribute of the upgrade state based on which flags we flip here. If we run through this multiple times and the first iteration unsets all the flags, I'd expect the follow up iterations to not have anything to set and we set the `mon_flags` attribute to an empty list. It looks like we're then using that attribute to set the flags again at the end. We might have to do this in the mgr module as part of the `upgrade_start` function instead and then set the mon flag attributes once in the initial upgrade state object created there so that we don't lose what we set during the upgrade

adk3798 reviewed (commented):
General structure LGTM, a couple specific comments

adk3798 reviewed (approved):
I would prefer the utility functions to set and check global mon flags lived in upgrade.py, but otherwise nothing worth blocking over here

rkachach commented:
@shraddhaag can you rebase plz
